onmt/transforms/terminology.py

```diff
@@ -0,0 +1,346 @@
+            if "'" not in key and "’" not in key and "‘" not in key
+        }
+        self.internal_termbase = self._create_internal_termbase(termbase_path)
+        self.automaton = self._create_automaton()
+        self.term_corpus_ratio = term_corpus_ratio
+        self.src_term_stoken = src_term_stoken
+        self.tgt_term_stoken = tgt_term_stoken
+        self.tgt_term_etoken = tgt_term_etoken
+        self.delimiter = delimiter
+
+    def _create_internal_termbase(self, termbase_path):
+        logger.debug("Creating termbase with lemmas for Terminology transform")
+
+        # Use Spacy's stopwords to get rid of junk entries
+        src_stopwords = self.src_nlp.Defaults.stop_words
+        tgt_stopwords = self.tgt_nlp.Defaults.stop_words
+        termbase = list()
+        with open(termbase_path, mode="r", encoding="utf-8") as file:
+            pairs = file.readlines()
+            for pair in pairs:
+                src_term, tgt_term = map(str, pair.split("\t"))
+                src_lemma = " ".join(
+                    "∥".join(tok.lemma_.split()) for tok in self.src_nlp(src_term)
+                ).strip()
+                tgt_lemma = " ".join(
+                    tok.lemma_ for tok in self.tgt_nlp(tgt_term)
+                ).strip()
+                if (
+                    src_lemma.lower() not in src_stopwords
+                    and tgt_lemma.lower() not in tgt_stopwords
+                ):
+                    termbase.append((src_lemma, tgt_lemma))
+        logger.debug(
+            f"Created termbase with {len(termbase)} lemmas "
+            f"for Terminology transform"
+        )
+        return termbase
+
+    def _create_automaton(self):
+        automaton = ahocorasick.Automaton()
+        for term in self.internal_termbase:
+            automaton.add_word(term[0], (term[0], term[1]))
+        automaton.make_automaton()
+        return automaton
+
+    def _src_sentence_with_terms(self, source_string, target_string) -> tuple:
+
+        maybe_augmented = source_string.split(self.delimiter)
+        source_only = maybe_augmented[0].strip()
+        augmented_part = (
+            maybe_augmented[1].strip() if len(maybe_augmented) > 1 else None
+        )
+
+        doc_src = self.src_nlp(source_only)
+        doc_tgt = self.tgt_nlp(target_string)
+
+        # Perform tokenization with spacy for consistency.
+        tokenized_source = [tok.text for tok in doc_src]
+        lemmatized_source = ["∥".join(tok.lemma_.lower().split()) for tok in doc_src]
+        lemmatized_target = [tok.lemma_.lower() for tok in doc_tgt]
+
+        lemmatized_source_string = " ".join(lemmatized_source)
+
+        offset = 0
+        source_with_terms = list()
+        term_counter = 0
+
+        max_terms_allowed = int(len(tokenized_source) * self.term_example_ratio)
+        is_match = False
+        for match_end, pair in self.automaton.iter_long(lemmatized_source_string):
```

Comment:
"pair" is not super understandable. We could naively think about it meaning "language pair".
Maybe something like (src_term, tgt_term) instead?

Reply:
If it's ok, I'll rename them to `(src_entry, tgt_entry)` as `src_term` and `tgt_term` are used later in the same method.